File: sdks/csharp/src/Croupier.Sdk/OpenAPIImporter.cs
// Copyright 2025 Croupier Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System.IO;
using System.Text;
using System.Text.Json;
using Croupier.Sdk.Models;

namespace Croupier.Sdk;

/// <summary>
/// OpenAPI 3 导入配置（与 Go SDK 的 ImportOptions 对齐）。
/// </summary>
public sealed class OpenAPIImportOptions
{
    /// <summary>为每个导入的 resource 添加前缀（如 "game"）。</summary>
    public string ResourcePrefix { get; set; } = string.Empty;

    /// <summary>为每个导入的 tag 添加前缀。</summary>
    public string TagPrefix { get; set; } = string.Empty;

    /// <summary>单个函数失败时是否继续导入其余函数。</summary>
    public bool ContinueOnError { get; set; }
}

/// <summary>
/// OpenAPI 3 导入助手，等价于 Go SDK 的 RegisterFromOpenAPI：
/// 解析 OpenAPI 3 文档，将每个 operation 转换为 FunctionDescriptor 并连同
/// 调用方提供的 handler 注册到 <see cref="CroupierClient"/>。
/// </summary>
public static class OpenAPIImporter
{
    private static readonly string[] OperationMethods =
    {
        "get", "put", "post", "delete", "options", "head", "patch", "trace",
    };

    /// <summary>
    /// 导入 OpenAPI 3 文档，将每个 operation 注册到 <paramref name="client"/>。
    /// </summary>
    /// <param name="client">目标 SDK 客户端。</param>
    /// <param name="specJson">OpenAPI 3 JSON 文档。</param>
    /// <param name="options">导入配置；可为 null。</param>
    /// <param name="handlerResolver">为派生函数 ID 提供 handler；返回 null 表示未提供。</param>
    /// <returns>已注册的函数 ID 列表。</returns>
    /// <exception cref="ArgumentException">spec 非法。</exception>
    /// <exception cref="InvalidOperationException">缺少 handler（未启用 ContinueOnError 时）。</exception>
    public static IReadOnlyList<string> RegisterFromOpenAPI(
        CroupierClient client,
        string specJson,
        OpenAPIImportOptions? options,
        Func<string, FunctionHandlerDelegate?> handlerResolver)
    {
        ArgumentNullException.ThrowIfNull(client);
        ArgumentNullException.ThrowIfNull(specJson);
        ArgumentNullException.ThrowIfNull(handlerResolver);

        JsonElement root;
        try
        {
            using var document = JsonDocument.Parse(specJson);
            root = document.RootElement.Clone();
        }
        catch (JsonException exception)
        {
            throw new ArgumentException($"load OpenAPI spec failed: {exception.Message}", exception);
        }
        if (root.ValueKind != JsonValueKind.Object
            || !root.TryGetProperty("paths", out var paths)
            || paths.ValueKind != JsonValueKind.Object)
        {
            throw new ArgumentException("OpenAPI spec must be an object containing 'paths'");
        }

        var registered = new List<string>();
        foreach (var pathEntry in paths.EnumerateObject())
        {
            if (pathEntry.Value.ValueKind != JsonValueKind.Object)
            {
                continue;
            }
            foreach (var method in OperationMethods)
            {
                if (!pathEntry.Value.TryGetProperty(method, out var operation)
                    || operation.ValueKind != JsonValueKind.Object)
                {
                    continue;
                }

                var descriptor = OperationToDescriptor(pathEntry.Name, operation, options);
                var handler = handlerResolver(descriptor.Id);
                if (handler is null)
                {
                    if (options is { ContinueOnError: true })
                    {
                        continue;
                    }
                    throw new InvalidOperationException($"no handler provided for function: {descriptor.Id}");
                }
                try
                {
                    client.RegisterFunction(descriptor, handler);
                }
                catch (Exception exception)
                {
                    if (options is { ContinueOnError: true })
                    {
                        continue;
                    }
                    throw new InvalidOperationException(
                        $"register function {descriptor.Id} failed: {exception.Message}", exception);
                }
                registered.Add(descriptor.Id);
            }
        }
        return registered;
    }

    /// <summary>
    /// 使用显式 handler 映射导入 OpenAPI 3 文档（等价 Go 的
    /// RegisterFromOpenAPIWithHandlers）。
    /// </summary>
    public static IReadOnlyList<string> RegisterFromOpenAPI(
        CroupierClient client,
        string specJson,
        OpenAPIImportOptions? options,
        IReadOnlyDictionary<string, FunctionHandlerDelegate> handlers)
    {
        ArgumentNullException.ThrowIfNull(handlers);
        return RegisterFromOpenAPI(client, specJson, options, id => handlers.TryGetValue(id, out var handler) ? handler : null);
    }

    internal static string DeriveOperationId(JsonElement operation, string path)
    {
        if (operation.TryGetProperty("operationId", out var operationId)
            && operationId.ValueKind == JsonValueKind.String
            && !string.IsNullOrEmpty(operationId.GetString()))
        {
            return operationId.GetString()!;
        }
        if (!string.IsNullOrEmpty(path))
        {
            var segments = path.Split('/', StringSplitOptions.RemoveEmptyEntries);
            if (segments.Length > 0)
            {
                return string.Join(".", segments);
            }
        }
        return "unknown.function";
    }

    internal static string DeriveSummary(JsonElement operation, string functionId)
    {
        if (operation.TryGetProperty("summary", out var summary)
            && summary.ValueKind == JsonValueKind.String
            && !string.IsNullOrEmpty(summary.GetString()))
        {
            return summary.GetString()!;
        }
        return functionId != "unknown.function" ? ToTitleCase(functionId) : "Unnamed Function";
    }

    internal static string ToTitleCase(string value)
    {
        var words = value.Split('_');
        for (var i = 0; i < words.Length; i++)
        {
            if (words[i].Length > 0)
            {
                words[i] = char.ToUpperInvariant(words[i][0]) + words[i][1..].ToLowerInvariant();
            }
        }
        return string.Join(' ', words);
    }

    internal static string ExtractExtension(JsonElement operation, string key)
    {
        if (!operation.TryGetProperty(key, out var value))
        {
            return string.Empty;
        }
        return value.ValueKind switch
        {
            JsonValueKind.String => value.GetString() ?? string.Empty,
            JsonValueKind.True => "true",
            JsonValueKind.False => "false",
            _ => value.GetRawText(),
        };
    }

    internal static string ParseRiskLevel(string level)
    {
        var normalized = level.ToLowerInvariant();
        return normalized switch
        {
            "low" or "safe" => "low",
            "high" => "high",
            "danger" or "critical" => "danger",
            _ => "medium",
        };
    }

    private static FunctionDescriptor OperationToDescriptor(
        string path, JsonElement operation, OpenAPIImportOptions? options)
    {
        var functionId = DeriveOperationId(operation, path);
        var descriptor = new FunctionDescriptor
        {
            Id = functionId,
            Summary = DeriveSummary(operation, functionId),
            Tags = new List<string>(),
        };

        if (operation.TryGetProperty("description", out var description)
            && description.ValueKind == JsonValueKind.String)
        {
            descriptor.Description = description.GetString();
        }
        if (operation.TryGetProperty("tags", out var tags) && tags.ValueKind == JsonValueKind.Array)
        {
            foreach (var tag in tags.EnumerateArray())
            {
                if (tag.ValueKind == JsonValueKind.String)
                {
                    descriptor.Tags.Add(tag.GetString()!);
                }
            }
        }

        var resource = ExtractExtension(operation, "x-resource");
        if (resource.Length > 0)
        {
            descriptor.Resource = resource;
        }
        var operationName = ExtractExtension(operation, "x-operation");
        if (operationName.Length > 0)
        {
            descriptor.Operation = operationName;
        }
        var permission = ExtractExtension(operation, "x-permission");
        if (permission.Length > 0)
        {
            descriptor.Permission = permission;
        }

        if (operation.TryGetProperty("requestBody", out var requestBody))
        {
            descriptor.InputSchema = JsonContentSchema(requestBody);
        }
        if (operation.TryGetProperty("responses", out var responses)
            && responses.ValueKind == JsonValueKind.Object
            && responses.TryGetProperty("200", out var ok))
        {
            descriptor.OutputSchema = JsonContentSchema(ok);
        }

        var risk = ExtractExtension(operation, "x-risk");
        descriptor.Risk = risk.Length == 0 ? "medium" : ParseRiskLevel(risk);

        if (options is not null)
        {
            if (options.ResourcePrefix.Length > 0 && descriptor.Resource is { Length: > 0 } target)
            {
                descriptor.Resource = $"{options.ResourcePrefix}.{target}";
            }
            if (options.TagPrefix.Length > 0)
            {
                descriptor.Tags = descriptor.Tags!.Select(tag => options.TagPrefix + tag).ToList();
            }
        }
        return descriptor;
    }

    /// <summary>浅层 OpenAPI schema → JSON Schema 转换（与 Go 对齐）。</summary>
    private static string? JsonContentSchema(JsonElement holder)
    {
        if (holder.ValueKind != JsonValueKind.Object
            || !holder.TryGetProperty("content", out var content)
            || content.ValueKind != JsonValueKind.Object
            || !content.TryGetProperty("application/json", out var media)
            || media.ValueKind != JsonValueKind.Object
            || !media.TryGetProperty("schema", out var schema)
            || schema.ValueKind != JsonValueKind.Object
            || !schema.EnumerateObject().Any())
        {
            return null;
        }

        using var buffer = new MemoryStream();
        using (var writer = new Utf8JsonWriter(buffer))
        {
            WriteSimplifiedSchema(writer, schema);
        }
        return Encoding.UTF8.GetString(buffer.ToArray());
    }

    private static void WriteSimplifiedSchema(Utf8JsonWriter writer, JsonElement schema)
    {
        writer.WriteStartObject();
        if (schema.TryGetProperty("type", out var type) && type.ValueKind == JsonValueKind.String)
        {
            writer.WriteString("type", type.GetString());
        }
        if (schema.TryGetProperty("description", out var description)
            && description.ValueKind == JsonValueKind.String
            && !string.IsNullOrEmpty(description.GetString()))
        {
            writer.WriteString("description", description.GetString());
        }
        if (schema.TryGetProperty("properties", out var properties) && properties.ValueKind == JsonValueKind.Object)
        {
            writer.WritePropertyName("properties");
            writer.WriteStartObject();
            foreach (var property in properties.EnumerateObject())
            {
                if (property.Value.ValueKind != JsonValueKind.Object)
                {
                    continue;
                }
                writer.WritePropertyName(property.Name);
                writer.WriteStartObject();
                var propertyType = property.Value.TryGetProperty("type", out var pt) && pt.ValueKind == JsonValueKind.String
                    ? pt.GetString()
                    : "object";
                writer.WriteString("type", propertyType);
                if (property.Value.TryGetProperty("description", out var pd)
                    && pd.ValueKind == JsonValueKind.String
                    && !string.IsNullOrEmpty(pd.GetString()))
                {
                    writer.WriteString("description", pd.GetString());
                }
                writer.WriteEndObject();
            }
            writer.WriteEndObject();
        }
        if (schema.TryGetProperty("required", out var required) && required.ValueKind == JsonValueKind.Array)
        {
            var items = required.EnumerateArray().Where(item => item.ValueKind == JsonValueKind.String).ToList();
            if (items.Count > 0)
            {
                writer.WritePropertyName("required");
                writer.WriteStartArray();
                foreach (var item in items)
                {
                    writer.WriteStringValue(item.GetString());
                }
                writer.WriteEndArray();
            }
        }
        writer.WriteEndObject();
    }
}

File: sdks/java/src/main/java/io/github/cuihairu/croupier/sdk/OpenAPIImporter.java
package io.github.cuihairu.croupier.sdk;

import io.github.cuihairu.croupier.sdk.invoker.Json;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.function.Function;

/**
 * OpenAPI 3 import helper mirroring the Go SDK's
 * {@code function.RegisterFromOpenAPI}: parses an OpenAPI 3 spec, converts
 * every operation into a {@link FunctionDescriptor} and registers it on a
 * {@link CroupierClient} with a caller-supplied handler.
 */
public final class OpenAPIImporter {

    private OpenAPIImporter() {
    }

    /** Controls OpenAPI import behaviour (mirrors Go ImportOptions). */
    public static final class ImportOptions {
        /** Prefix prepended to every imported resource (e.g. "game"). */
        private String resourcePrefix = "";
        /** Prefix prepended to every imported tag. */
        private String tagPrefix = "";
        /** Default timeout in milliseconds (Go parity; the Java descriptor carries no timeout field yet). */
        private long defaultTimeoutMs = 0;
        /** Keep importing remaining operations when one fails. */
        private boolean continueOnError = false;

        public ImportOptions resourcePrefix(String resourcePrefix) {
            this.resourcePrefix = resourcePrefix == null ? "" : resourcePrefix;
            return this;
        }

        public ImportOptions tagPrefix(String tagPrefix) {
            this.tagPrefix = tagPrefix == null ? "" : tagPrefix;
            return this;
        }

        public ImportOptions defaultTimeoutMs(long defaultTimeoutMs) {
            this.defaultTimeoutMs = defaultTimeoutMs;
            return this;
        }

        public ImportOptions continueOnError(boolean continueOnError) {
            this.continueOnError = continueOnError;
            return this;
        }
    }

    private static final List<String> OPERATION_METHODS = List.of(
        "get", "put", "post", "delete", "options", "head", "patch", "trace");

    /**
     * Imports an OpenAPI 3 spec, registering every operation on {@code client}.
     *
     * @param client      target SDK client
     * @param spec        OpenAPI 3 JSON document
     * @param options     import options (may be {@code null})
     * @param resolver    supplies a handler for a derived function ID; returning
     *                    {@code null} marks the operation as unhandled
     * @return the list of registered function IDs
     * @throws CroupierException wrapping parse/registration failures
     */
    public static List<String> registerFromOpenAPI(
        CroupierClient client,
        String spec,
        ImportOptions options,
        Function<String, FunctionHandler> resolver) throws CroupierException {

        Object documentRaw;
        try {
            documentRaw = Json.parse(spec);
        } catch (IllegalArgumentException exception) {
            throw new CroupierException("load OpenAPI spec failed: " + exception.getMessage(), exception);
        }
        if (!(documentRaw instanceof Map<?, ?> document) || !(document.get("paths") instanceof Map<?, ?>)) {
            throw new CroupierException("OpenAPI spec must be an object containing 'paths'");
        }

        List<String> registered = new ArrayList<>();
        for (Map.Entry<?, ?> pathEntry : ((Map<?, ?>) document.get("paths")).entrySet()) {
            if (!(pathEntry.getKey() instanceof String path) || !(pathEntry.getValue() instanceof Map<?, ?> pathItem)) {
                continue;
            }
            for (String method : OPERATION_METHODS) {
                if (!(pathItem.get(method) instanceof Map<?, ?> operation)) {
                    continue;
                }
                @SuppressWarnings("unchecked")
                Map<String, Object> operationMap = (Map<String, Object>) operation;
                FunctionDescriptor descriptor = operationToDescriptor(path, operationMap, options);
                FunctionHandler handler = resolver.apply(descriptor.getId());
                if (handler == null) {
                    if (options != null && options.continueOnError) {
                        continue;
                    }
                    throw new CroupierException("no handler provided for function: " + descriptor.getId());
                }
                try {
                    client.registerFunction(descriptor, handler);
                } catch (Exception exception) {
                    if (options != null && options.continueOnError) {
                        continue;
                    }
                    throw new CroupierException("register function " + descriptor.getId() + " failed: " + exception.getMessage(), exception);
                }
                registered.add(descriptor.getId());
            }
        }
        return registered;
    }

    /**
     * Imports an OpenAPI 3 spec using an explicit handler map
     * (mirrors Go RegisterFromOpenAPIWithHandlers).
     */
    public static List<String> registerFromOpenAPIWithHandlers(
        CroupierClient client,
        String spec,
        ImportOptions options,
        Map<String, FunctionHandler> handlers) throws CroupierException {
        return registerFromOpenAPI(client, spec, options, handlers::get);
    }

    // ------------------------------------------------------------------

    @SuppressWarnings("unchecked")
    private static FunctionDescriptor operationToDescriptor(
        String path, Map<String, Object> operation, ImportOptions options) {

        String functionId = deriveOperationId(operation, path);
        FunctionDescriptor descriptor = new FunctionDescriptor();
        descriptor.setId(functionId);
        descriptor.setSummary(deriveSummary(operation, functionId));
        descriptor.setDescription(stringOrNull(operation.get("description")));
        descriptor.setTags(stringTags(operation.get("tags")));
        descriptor.setResource(emptyToNull(extractExtension(operation, "x-resource")));
        descriptor.setOperation(emptyToNull(extractExtension(operation, "x-operation")));
        descriptor.setPermission(emptyToNull(extractExtension(operation, "x-permission")));
        descriptor.setCapability(emptyToNull(extractExtension(operation, "x-capability")));
        descriptor.setExecution(emptyToNull(extractExtension(operation, "x-execution")));
        applyApprovalExtension(descriptor, operation.get("x-approval"));

        String inputSchema = jsonContentSchema(operation.get("requestBody"));
        if (inputSchema != null) {
            descriptor.setInputSchema(inputSchema);
        }
        if (operation.get("responses") instanceof Map<?, ?> responses
            && responses.get("200") instanceof Map<?, ?> ok) {
            String outputSchema = jsonContentSchema(ok);
            if (outputSchema != null) {
                descriptor.setOutputSchema(outputSchema);
            }
        }

        String risk = extractExtension(operation, "x-risk");
        descriptor.setRisk(risk.isEmpty() ? "warning" : parseRiskLevel(risk));

        if (options != null) {
            if (!options.resourcePrefix.isEmpty() && descriptor.getResource() != null) {
                descriptor.setResource(options.resourcePrefix + "." + descriptor.getResource());
            }
            if (!options.tagPrefix.isEmpty()) {
                List<String> prefixed = new ArrayList<>();
                for (String tag : descriptor.getTags()) {
                    prefixed.add(options.tagPrefix + tag);
                }
                descriptor.setTags(prefixed);
            }
        }
        return descriptor;
    }

    static String deriveOperationId(Map<String, Object> operation, String path) {
        String operationId = stringOrNull(operation.get("operationId"));
        if (operationId != null && !operationId.isEmpty()) {
            return operationId;
        }
        if (path != null && !path.isEmpty()) {
            List<String> segments = new ArrayList<>();
            for (String segment : path.split("/")) {
                if (!segment.isEmpty()) {
                    segments.add(segment);
                }
            }
            if (!segments.isEmpty()) {
                return String.join(".", segments);
            }
        }
        return "unknown.function";
    }

    static String deriveSummary(Map<String, Object> operation, String functionId) {
        String summary = stringOrNull(operation.get("summary"));
        if (summary != null && !summary.isEmpty()) {
            return summary;
        }
        if (functionId != null && !"unknown.function".equals(functionId)) {
            return toTitleCase(functionId);
        }
        return "Unnamed Function";
    }

    static String toTitleCase(String value) {
        String[] words = value.split("_");
        StringBuilder output = new StringBuilder();
        for (int i = 0; i < words.length; i++) {
            if (i > 0) {
                output.append(' ');
            }
            String word = words[i];
            if (!word.isEmpty()) {
                output.append(Character.toUpperCase(word.charAt(0)));
                if (word.length() > 1) {
                    output.append(word.substring(1).toLowerCase(Locale.ROOT));
                }
            }
        }
        return output.toString();
    }

    /** Shallow OpenAPI-schema -> JSON-Schema conversion (Go parity). */
    @SuppressWarnings("unchecked")
    private static String jsonContentSchema(Object holder) {
        if (!(holder instanceof Map<?, ?> holderMap) || !(holderMap.get("content") instanceof Map<?, ?> content)) {
            return null;
        }
        if (!(content.get("application/json") instanceof Map<?, ?> media) || !(media.get("schema") instanceof Map<?, ?> schema)) {
            return null;
        }
        Map<String, Object> schemaMap = (Map<String, Object>) schema;
        if (schemaMap.isEmpty()) {
            return null;
        }
        Map<String, Object> result = new java.util.LinkedHashMap<>();
        String type = stringOrNull(schemaMap.get("type"));
        if (type != null && !type.isEmpty()) {
            result.put("type", type);
        }
        String description = stringOrNull(schemaMap.get("description"));
        if (description != null && !description.isEmpty()) {
            result.put("description", description);
        }
        if (schemaMap.get("properties") instanceof Map<?, ?> properties) {
            Map<String, Object> props = new java.util.LinkedHashMap<>();
            for (Map.Entry<?, ?> entry : properties.entrySet()) {
                if (entry.getKey() instanceof String name && entry.getValue() instanceof Map<?, ?> rawProp) {
                    Map<String, Object> entry_ = new java.util.LinkedHashMap<>();
                    String propType = stringOrNull(((Map<String, Object>) rawProp).get("type"));
                    entry_.put("type", propType == null || propType.isEmpty() ? "object" : propType);
                    String propDescription = stringOrNull(((Map<String, Object>) rawProp).get("description"));
                    if (propDescription != null && !propDescription.isEmpty()) {
                        entry_.put("description", propDescription);
                    }
                    props.put(name, entry_);
                }
            }
            result.put("properties", props);
        }
        if (schemaMap.get("required") instanceof List<?> required && !required.isEmpty()) {
            result.put("required", required);
        }
        if (result.isEmpty()) {
            return null;
        }
        return Json.stringify(result);
    }

    static String extractExtension(Map<String, Object> operation, String key) {
        Object value = operation.get(key);
        if (value == null) {
            return "";
        }
        if (value instanceof String text) {
            return text;
        }
        if (value instanceof Boolean flag) {
            return flag ? "true" : "false";
        }
        return Json.stringify(value);
    }

    static String parseRiskLevel(String level) {
        String normalized = level.toLowerCase(Locale.ROOT);
        switch (normalized) {
            case "low":
            case "safe":
                return "safe";
            case "medium":
            case "moderate":
            case "warning":
                return "warning";
            case "high":
                return "high";
            case "danger":
            case "critical":
                return "danger";
            default:
                return "warning";
        }
    }

    /** Maps {@code x-approval: {required, policyKey}} onto descriptor v2 fields. */
    private static void applyApprovalExtension(FunctionDescriptor descriptor, Object approval) {
        if (!(approval instanceof Map<?, ?> approvalMap)) {
            return;
        }
        Object required = approvalMap.get("required");
        if (required instanceof Boolean flag) {
            descriptor.setApprovalRequired(flag);
        } else if (required instanceof String text && !text.isEmpty()) {
            descriptor.setApprovalRequired(Boolean.parseBoolean(text));
        }
        if (approvalMap.get("policyKey") instanceof String policyKey && !policyKey.isEmpty()) {
            descriptor.setApprovalPolicyKey(policyKey);
        }
    }

    private static String stringOrNull(Object value) {
        return value instanceof String text ? text : null;
    }

    private static String emptyToNull(String value) {
        return value == null || value.isEmpty() ? null : value;
    }

    private static List<String> stringTags(Object value) {
        List<String> tags = new ArrayList<>();
        if (value instanceof List<?> list) {
            for (Object item : list) {
                if (item instanceof String tag) {
                    tags.add(tag);
                }
            }
        }
        return tags;
    }
}
